src/condor_utils/file_transfer.cpp

```diff
@@ -2150,28 +2151,125 @@ FileTransfer::AddDownloadFilenameRemaps(char const *remaps) {
+	std::filesystem::path partial_path( prefix );
+	auto next_path_component_iter = suffix.begin();
+	while( std::filesystem::exists( partial_path ) ) {
+		// dprintf( D_ZKM, "shadow_safe_mkdir_impl(%s, %s), (1): partial_path = %s, next_path_component_iter = %s\n", prefix.c_str(), suffix.c_str(), partial_path.c_str(), next_path_component_iter->c_str() );
+		if( next_path_component_iter == suffix.end() ) {
+			// Then the directory we're trying to create exists.  Great!
+			// (This should only happen if somebody else has created it
+			// between when we checked in shadow_safe_mkdir() and here.)
+			// We don't check allow_shadow_access() here because we're not
+			// writing anything now, and it'll be called for each individual
+			// file write anyway.
+			return true;
 		}
 
-		idx = buffer.rfind( DIR_DELIM_CHAR );
-		if (idx != std::string::npos) {
-			buffer.erase(idx);
-		}
-	} while( idx != std::string::npos );
+		partial_path = partial_path / (* next_path_component_iter);
+		++next_path_component_iter;
+	}
+
+	// Step (2A).
+	// dprintf( D_ZKM, "shadow_safe_mkdir_impl(%s, %s), (2A): partial_path = %s, next_path_component_iter = %s\n", prefix.c_str(), suffix.c_str(), partial_path.c_str(), next_path_component_iter->c_str() );
+	if(! allow_shadow_access( partial_path.c_str() )) {
+		return false;
+	}
+
+	// Step (2B).  We can see EEXIST if, for instance, some other shadow
+	// is going through this logic at the same time.  In general, I'm not
+	// going to worry if we get ENOENT; if some other process is deleting
+	// directories where you want to write your output data, I think it's
+	// entirely appropriate for us to fail to write there.
+	int rv = mkdir( partial_path.c_str(), mode );
+	if( rv != 0 && errno != EEXIST ) {
+		return false;
+	}
+
+	// Step (2C).
+	// dprintf( D_ZKM, "shadow_safe_mkdir_impl(%s, %s), (2C): partial_path = %s, next_path_component_iter = %s\n", prefix.c_str(), suffix.c_str(), partial_path.c_str(), next_path_component_iter->c_str() );
+	if( next_path_component_iter == suffix.end() ) { return true; }
+
+	// Step (2D).
+	std::filesystem::path remainder;
+	for( ; next_path_component_iter != suffix.end(); ++next_path_component_iter ) {
+		remainder /= (* next_path_component_iter);
+	}
+
+	// dprintf( D_ZKM, "shadow_safe_mkdir_impl(%s, %s), (2D): partial_path = %s, remainder = %s\n", prefix.c_str(), suffix.c_str(), partial_path.c_str(), remainder.c_str() );
+	return shadow_safe_mkdir_impl( partial_path, remainder, mode );
 }
 
 
+bool
+shadow_safe_mkdir( const std::string & dir, mode_t mode, priv_state priv ) {
+	bool rv;
+	priv_state saved_priv;
+
+	std::filesystem::path path(dir);
+	if(! path.has_root_path()) {
+		dprintf( D_ALWAYS, "Internal logic error: shadow_safe_mkdir() called with relative path.  Refusing to make the directory.\n" );
+		return false;
+	}
+
+	// If the path already exists, we have nothing to do.
+	if( std::filesystem::exists( path ) ) { return true; }
+
+
+	if( priv != PRIV_UNKNOWN ) {
```

Comment:
The priv switch needs to happen before the exists() call.
TemporaryPrivSentry is how the cool kids ensure we restore the original priv state by the end of the function.

Reply:
New commit makes these changes.